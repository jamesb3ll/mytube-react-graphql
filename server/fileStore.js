import fs from 'fs';

export const UPLOAD_DIR = __dirname + '/uploads';

export function loadVideos() {
  return fs
    .readdirSync(UPLOAD_DIR)
    .filter(file => !file.startsWith('.'))
    .map((file, index) => {
      // Sorry... No DB for video names
      const [hash, name] = file.split('-');
      return {
        name: name || `Video ${index + 1}`,
        file: `${UPLOAD_DIR}/${file}`,
      };
    });
}

export function saveVideo(filename, stream) {
  // naive random id implementation
  const id = Math.random()
    .toString(36)
    .slice(2);
  const path = `${UPLOAD_DIR}/${id}-${filename}`;
  return new Promise((resolve, reject) =>
    stream
      .pipe(fs.createWriteStream(path))
      .on('error', error => reject(error))
      .on('finish', () => resolve({ path }))
  );
}
