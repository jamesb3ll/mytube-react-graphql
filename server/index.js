// node --experimental-modules doesn't support importing named exports from a commonjs module
// esm supports ECMAScript modules in Node 6+ (https://github.com/standard-things/esm)
// eslint-disable-next-line no-native-reassign
require = require('esm')(module);
module.exports = require('./server');
