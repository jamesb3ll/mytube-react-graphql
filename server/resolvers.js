import { GraphQLUpload } from 'apollo-server';

export default {
  Upload: GraphQLUpload,
  Query: {
    videos(parent, args, { videos }) {
      return videos;
    },
  },
  Mutation: {
    async uploadVideo(parent, { name, file }, { videos, saveVideo }) {
      const { filename, mimetype, createReadStream } = await file;

      // Naive video format verification
      if (!mimetype.includes('video')) {
        throw new Error('Only video formats allowed.');
      }

      const stream = createReadStream();
      const { path } = await saveVideo(filename, stream);
      videos.push({ name, file: path });

      return { name, file: path };
    },
  },
};
