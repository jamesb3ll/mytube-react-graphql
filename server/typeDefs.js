import { gql } from 'apollo-server';

export default gql`
  type Video {
    name: String!
    file: String!
  }
  type Query {
    videos: [Video]
  }
  type Mutation {
    uploadVideo(name: String!, file: Upload!): Video!
  }
`;
