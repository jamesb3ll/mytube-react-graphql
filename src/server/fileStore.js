import fs from 'fs';

export const UPLOAD_DIR = __dirname + '/uploads/';

export function loadVideos() {
  return fs
    .readdirSync(UPLOAD_DIR)
    .filter(file => !file.startsWith('.'))
    .map((file, index) => {
      // Sorry... No DB for storing video names
      const [id, name] = file.split('-');
      return {
        name: name || `Video ${index + 1}`,
        file,
      };
    });
}

export function saveVideo(filename, stream) {
  // naive random id implementation
  const id = Math.random()
    .toString(36)
    .slice(2);
  const file = `${id}-${filename}`;
  const path = `${UPLOAD_DIR}/${file}`;
  return new Promise((resolve, reject) =>
    stream
      .pipe(fs.createWriteStream(path))
      .on('error', error => reject(error))
      .on('finish', () => resolve({ path, file }))
  );
}
