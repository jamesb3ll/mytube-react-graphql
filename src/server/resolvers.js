import { GraphQLUpload } from 'apollo-server-express';

export default {
  Upload: GraphQLUpload,
  Query: {
    videos(parent, args, { videos }) {
      return videos;
    },
  },
  Mutation: {
    async uploadVideo(parent, { name, file: upload }, { videos, saveVideo }) {
      const { filename, mimetype, createReadStream } = await upload;

      // Naive video format verification
      if (!mimetype.includes('video')) {
        throw new Error('Only video formats allowed.');
      }

      const stream = createReadStream();
      const { file } = await saveVideo(filename, stream);
      videos.push({ name, file });

      return { name, file };
    },
  },
};
