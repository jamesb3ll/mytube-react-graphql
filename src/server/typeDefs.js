import { gql } from 'apollo-server-express';

export default gql`
  type Video {
    name: String!
    file: String!
  }
  type Query {
    videos: [Video]
  }
  type Mutation {
    uploadVideo(name: String!, file: Upload!): Video!
  }
`;
