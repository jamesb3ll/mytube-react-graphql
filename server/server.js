import { ApolloServer } from 'apollo-server';
import typeDefs from './typeDefs';
import resolvers from './resolvers';
import { loadVideos, saveVideo } from './fileStore';

// Load video files when server boots and keep in-memory store
const videos = loadVideos();

export const server = new ApolloServer({
  typeDefs,
  resolvers,
  context: {
    videos,
    saveVideo,
  },
  uploads: {
    maxFileSize: 10000000, // 10 MB
    maxFiles: 1,
  },
});

server.listen({ port: 4000 }).then(({ url }) => {
  console.log(`ğŸš€  Server started at ${url}`);
});
