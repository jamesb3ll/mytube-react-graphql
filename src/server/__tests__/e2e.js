import { createTestClient } from 'apollo-server-testing';
import { server } from '../server';
import { gql } from 'apollo-server-express';
import { createReadStream } from 'fs';

const GET_VIDEOS = gql`
  query {
    videos {
      name
      file
    }
  }
`;

const UPLOAD_VIDEO = gql`
  mutation uploadVideo($name: String!, $file: Upload!) {
    uploadVideo(name: $name, file: $file) {
      name
      file
    }
  }
`;

describe('Server', () => {
  let query, mutate, stop;

  beforeEach(async () => {
    const testServer = await createTestClient(server);
    query = testServer.query;
    mutate = testServer.mutate;
    stop = testServer.stop;
  });

  afterAll(() => stop());

  it('gets videos', async () => {
    const res = await query({ query: GET_VIDEOS });

    expect(res).toMatchSnapshot();
  });

  // Skipping: Testing file uploads requires building the request with FormData
  // https://github.com/jaydenseric/graphql-upload/issues/125
  it.skip('uploads videos', async () => {
    const file = createReadStream('./mockFiles/surf.mp4');

    const res = await mutate({
      query: UPLOAD_VIDEO,
      variables: {
        name: 'Test',
        file,
      },
    });

    expect(res).toMatchSnapshot();
  });
});
