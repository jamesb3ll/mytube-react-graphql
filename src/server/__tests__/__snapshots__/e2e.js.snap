// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Server gets videos 1`] = `
Object {
  "data": Object {
    "videos": Array [
      Object {
        "file": "02rwt6csyud2-Surf.mp4",
        "name": "Surf.mp4",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
